docs/user/build-customization.rst

```diff
@@ -0,0 +1,155 @@
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       pre_install:
+         - echo "Executing pre_install step..."
+         - bash ./scripts/pre_install.sh
+       post_build:
+         - echo "Executing post_build step..."
+         - curl -X POST \
+           -F "project=${READTHEDOCS_PROJECT}" \
+           -F "version=${READTHEDOCS_VERSION}" https://my.company.com/webhooks/readthedocs/
+
+.. note::
+
+   Currently, the default jobs (e.g. ``checkout``, ``system_dependencies``, etc) executed by Read the Docs are not possible to overwritte or skip.
+
+
+Examples
+--------
+
+This is a non-exahustive list of common examples where using :ref:`config-file/v2:build.jobs` will be useful.
+Also note that these examples may require some adaption for each projects' use case.
+They are listed here just as a guide.
+
+
+Unshallow clone
+~~~~~~~~~~~~~~~
+
+Read the Docs does not perform a full clone on ``checkout`` step to reduce network data and speed up the build process.
+Because of this reason, if there are extensions in the build process that depend on the full history, they will fail.
+To avoid this, it's possible to unshallow the clone done by Read the Docs:
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       post_checkout:
+         - git fetch --unshallow
+
+
+Generate documentation from annotated sources with Doxygen
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It's possible to run Doxygen as part of the build process to generate documentation from annotated sources:
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       pre_build:
+         - doxygen
+
+
+Use MkDocs extensions with extra required steps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are some MkDocs extensions that require specific commands to be run to generate extra pages before performing the build.
+For example, `pydoc-markdown <http://niklasrosenstein.github.io/pydoc-markdown/>`_
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       pre_build:
+         - pydoc-markdown --build --site-dir "$PWD/_build/html"
+
+
+Avoid having a dirty ``git`` index
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Read the Docs needs to modify some files before performing the build to be able to integrate with some of its features.
+Because of this reason, it could happen the git index gets dirty (e.i. it will detect modified files).
+In case this happens and the project is using any kind of extension that generates a version based on git metadata (like `setuptools_scm <https://github.com/pypa/setuptools_scm/>`_),
+this could case a miss generated version number.
+In that case, the git index can be updated to ignore the files that Read the Docs has modified.
```

Comment:
```suggestion
Read the Docs needs to modify some files before performing the build to be able to integrate with some of its features.
Because of this reason, it could happen the git index gets dirty (it will detect modified files).
In case this happens and the project is using any kind of extension that generates a version based on git metadata (like `setuptools_scm <https://github.com/pypa/setuptools_scm/>`_),
this could cause an invalid version number to be generated.
In that case, the git index can be updated to ignore the files that Read the Docs has modified.
```